src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs

```diff
@@ -35,6 +35,7 @@ internal abstract class SourceOrdinaryMethodSymbolBase : SourceOrdinaryMethodOrU
             bool isIterator,
             bool isExtensionMethod,
             bool isPartial,
+            bool isReadOnly,
```

Comment:
I think this is not the right way to set `isReadOnly` on the synthesized method. Instead of adding a constructor parameter I think we should adjust the value returned by `protected override DeclarationModifiers MakeDeclarationModifiers` in the appropriate synthesized method symbols, e.g. `SynthesizedRecordEquals.MakeDeclarationModifiers`.

Reply:
I'm looking a little closer at the existing `isPartial` and how it is used. It might be reasonable in this design for `SourceOrdinaryMethodSymbol` to check the modifiers in a similar way when passing an argument for it to `SourceOrdinaryMethodSymbolBase`. I will follow up when I'm a little more confident in what the right way is here.

Reply:
My initial approach was to change `MakeDeclarationModifiers` to add `readonly` declaration. You can see it here  https://github.com/dotnet/roslyn/pull/55039/files/77be45707a517f0684370ae072a20150f488cfcb
The problem with that is `MakeDeclarationModifiers` being a virtual method still called in the constructor of the base class, so for example in the case with `Deconstruct` `positionalMembers` field are not initialized at the moment of execution.
And as you point out there is a similar problem with `PrintMembers` because we need information about record members.

Would renaming the flag into `IsSynthesizedAsReadOnly` help here? Or should we search for a different approach completely? Check for marking as `readonly` need to be executed before the base constructor, so it does not allow a lot of options. And, I imagine, moving the evaluation of modifiers from the base constructor would be quite problematic as well.

Reply:
Thanks for the info. You're correct that my initial suggestion doesn't result in the correct semantics.

I think for now what I'd like to do is add an additional assertion to `SourceOrdinaryMethodSymbolBase` to make it clear that this isn't how explicitly declared methods are made `readonly`:

```cs
Debug.Assert(!isReadOnly || this.IsImplicitlyDeclared,
    "We only expect synthesized methods to use this flag to make a method readonly. Explicitly declared methods should get this value from modifiers in syntax.");
```

I think in the future I'd like to make an attempt to push the virtual call out of the constructor, perhaps into the constructor arguments. But this PR doesn't seem like the right time to make such an attempt.

Reply:
Thanks, I've added assert. Do you mind if I try to move the virtual call out after this PR?

Reply:
Feel free to give it a try.